src/tabs/console/ConsoleAutocomplete.tsx

```diff
@@ -0,0 +1,70 @@
+// Tab completion engine for autocompleting commands. Inspired by the TabCompletionEngine class
+// for the Terasology in-game console.
+
+export class ConsoleAutocomplete {
+
+  // autocomplete the command.
+  public static complete(command: string): string {
+    this.previousCommands.push(command);
+    // if command was not entered, display some help text
+    if (command.length <= 0) {
+      return "Type 'help' to list all commands.";
+    }
+    // if no commands were previously entered, find a list of commands
+    if (this.previousCommands.length === 1) {
+      this.findMatches(command);
+    }
+    // if matches were found, do nothing to the command
+    if (this.matches.length === 0) {
+      return command;
+    } else if (this.matches.length === 1) {
+      return this.matches[0];
+    }
+    // If the last possible match has already been iterated through, reset the index to the first location.
+    // The match index must be offset because the length function returns 1 with one element, but its index
+    // is zero.
+    if (this.matchIndex + 1 === this.matches.length) {
+      this.matchIndex = -1;
+    }
+    // If the first command entered had more than one match, cycle throught the matches.
+    // previousCommands is always at least size 1.
+    if (this.previousCommands.indexOf(this.previousCommands[0]) !== -1 && this.previousCommands.length !== 1) {
```

Comment:
I don't understand this line. If we let `X` be `this.previousCommands`, this could be rewritten as `X.indexOf(X[0])`. Which means "give me the index of item at index 0", which is always 0, correct? Thus the 1st condition is always `0 !== -1`, which is `true`. Also, `true && Y == Y` for any `Y` so you could just put the 2nd condition (`this.previousCommands.length !== 1`) alone.

Is my reasoning correct? Maybe I'm missing something. Also, the comment mentions matches, but the code only actually looks at previous commands. Could you help me understanding what this part does? I understand you have to increment the index until there are matches, but not how it's checked here

Reply:
When I wrote the first iteration of the code I was doing something like what the comment said. I later rewrote a part of it to fix issues that came up, and I guess I forgot to update the comment and code. You are right about the logic.